dart/dynamics/HierarchicalIK.h

```diff
@@ -0,0 +1,383 @@
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_DYNAMICS_HIERARCHICALIK_H_
+#define DART_DYNAMICS_HIERARCHICALIK_H_
+
+#include <unordered_set>
+
+#include "dart/dynamics/InverseKinematics.h"
+
+namespace dart {
+namespace dynamics {
+
+/// An IKHierarchy is a sorted set of IK modules. The outer vector represents
+/// the precedence of the IK modules. Modules with the same precedence will have
+/// their gradients added. Precedence of the modules decreases as the index of
+/// the outer vector increases. Modules with lower precedence will be projected
+/// through the null spaces of modules with higher precedence.
+typedef std::vector< std::vector< std::shared_ptr<InverseKinematics> > > IKHierarchy;
+
+/// The HierarchicalIK class provides a convenient way of setting up a
+/// hierarchical inverse kinematics optimization problem which combines several
+/// InverseKinematics problems into one. InverseKinematics problems with a
+/// larger hierarchy level will be projected into null spaces of the problems
+/// that have a smaller hierarchy number.
+///
+/// Note that the HierarchicalIK will only account for the
+/// InverseKinematics::ErrorMethod and InverseKinematics::GradientMethod that
+/// the IK modules specify; it will ignore any other constraints or objectives
+/// put into the IK modules' Problems. Any additional constraints or objectives
+/// that you want the HierarchicalIK to solve should be put directly into the
+/// HierarchicalIK's Problem.
+class HierarchicalIK : public common::Subject
+{
+public:
+
+  /// Virtual destructor
+  virtual ~HierarchicalIK() = default;
+
+  /// Solve the IK Problem. By default, the Skeleton itself will retain the
+  /// solved joint configuration. If you pass in true for _resetConfiguration,
+  /// then the joint positions will be return to their original configuration
+  /// after the problem is solved.
+  bool solve(bool _resetConfiguration = false);
```

Comment:
@mkoval might have an opinion on this since I believe he's already using this functionality in his code.

I actually agree with you that ``solve(true)`` looks more like it should apply the solution rather than reset the configuration; I just hadn't given it much consideration at the time.